### ISSUE TITLE
support class static blocks (stage-4)

### ISSUE THREAD

**Issue by aladdin-add** [eslint/eslint#15016]
**What is the name of the syntax to implement?**
class static blocks

**Please provide the TC39 URL for the syntax proposal:**
https://github.com/tc39/proposal-class-static-block

**Please provide some example code for the new syntax:**

```js
// with static blocks:
class C {
  static x = ...;
  static y;
  static z;
  static {
    try {
      const obj = doSomethingWith(this.x);
      this.y = obj.y;
      this.z = obj.z;
    }
    catch {
      this.y = ...;
      this.z = ...;
    }
  }
}
```

## Implementation Checklist

Please check off all items that have already been completed. Be sure to paste the pull request URLs next to each item so we can verify the work as done.

- [x] Ecma262 update: https://github.com/tc39/ecma262/pull/2440
- [x] ESTree update: https://github.com/estree/estree/pull/254
- [x] Acorn update: https://github.com/acornjs/acorn/pull/1063
- [x] `eslint-visitor-keys` update: https://github.com/eslint/eslint-visitor-keys/pull/29, https://github.com/eslint/eslint/pull/15277
- [x] `espree` update: https://github.com/eslint/espree/pull/518
- [x] `eslint-scope` update: https://github.com/eslint/eslint-scope/pull/80, https://github.com/eslint/eslint/pull/15321
- [x] `eslint` update:
  - [x] Code Path Analysis https://github.com/eslint/eslint/pull/15282
  - [x] Rules
    - [x] `block-scoped-var` https://github.com/eslint/eslint/pull/15302
    - [x] `block-spacing` https://github.com/eslint/eslint/pull/15297
    - [x] `brace-style` https://github.com/eslint/eslint/pull/15322
    - [x] `class-methods-use-this` https://github.com/eslint/eslint/pull/15298
    - [x] `complexity` https://github.com/eslint/eslint/pull/15328
    - [x] `indent` https://github.com/eslint/eslint/pull/15324
    - [x] `keyword-spacing` https://github.com/eslint/eslint/pull/15289
    - [x] `lines-around-comment` https://github.com/eslint/eslint/pull/15323
    - [x] `max-depth` https://github.com/eslint/eslint/pull/15316
    - [x] `max-statements` https://github.com/eslint/eslint/pull/15315
    - [x] `no-lone-blocks` https://github.com/eslint/eslint/pull/15295
    - [ ] ~`no-empty` (or `no-empty-function`?)~ neither, possibly a new rule.
    - [x] `no-eval` https://github.com/eslint/eslint/pull/15300
    - [x] `no-extra-semi` https://github.com/eslint/eslint/pull/15287
    - [x] `no-inner-declarations` https://github.com/eslint/eslint/pull/15290
    - [x] `no-invalid-this` https://github.com/eslint/eslint/pull/15300
    - [x] `no-redeclare` https://github.com/eslint/eslint/pull/15313
    - [x] `no-unused-expressions` https://github.com/eslint/eslint/pull/15283
    - [x] `no-use-before-define` https://github.com/eslint/eslint/pull/15312
    - [x] `one-var` https://github.com/eslint/eslint/pull/15317
    - [x] `padded-blocks` https://github.com/eslint/eslint/pull/15333
    - [x] `padding-line-between-statements` https://github.com/eslint/eslint/pull/15318
    - [x] `prefer-const` https://github.com/eslint/eslint/pull/15325
    - [x] `semi` https://github.com/eslint/eslint/pull/15286
    - [x] `semi-style` https://github.com/eslint/eslint/pull/15309
    - [x] `strict` https://github.com/eslint/eslint/pull/15314
    - [x] `vars-on-top` https://github.com/eslint/eslint/pull/15306

**Are you willing to submit a pull request to implement this syntax?**
y

**Comment by nzakas** [comment]
This looks like it will need a new scope in `eslint-scope` and a new code path type.

**Comment by mdjermanovic** [comment]
> This looks like it will need a new scope in `eslint-scope` and a new code path type.

Yes, I think the same.

**Comment by mdjermanovic** [comment]
I unchecked `espree` because it also re-exports visitor keys. I don't think we're using `espree.VisitorKeys` in ESLint, but `espree` itself would be incomplete without updating `eslint-visitor-keys` dependency.

**Comment by mdjermanovic** [comment]
I added an initial list of rules that should be updated. A couple of them may need only docs/tests update. Note that class static blocks do not have directive prologue.

**Comment by mdjermanovic** [comment]
> * [ ]  `no-empty` (or `no-empty-function`?)

@eslint/eslint-tsc thoughts about this? Some rules treat block statements and function bodies equally, but this is a case where one rule specifically targets real block statements, and the other rule targets function bodies. Shall we in this case (and generally) treat class static blocks as blocks or functions? They look more like functions to me. Here's an interesting discussion with opposite views: https://github.com/tc39/proposal-class-static-block/issues/7.

**Comment by ljharb** [comment]
They have a scope, and aren’t simply a block, but they explicitly aren’t functions and don’t add a stack frame.

**Comment by mdjermanovic** [comment]
> and don’t add a stack frame

By the [proposal](https://github.com/tc39/proposal-class-static-block), they should?

> A `static {}` initialization block should be represented as an independent stack frame in debuggers and exception traces.

**Comment by mdjermanovic** [comment]
And each static block has a function object that is called during the class definition evaluation:

https://tc39.es/ecma262/#sec-classstaticblockdefinition-record-specification-type

**Comment by ljharb** [comment]
Oof, ok, that’s interesting. I don’t recall discussing that in plenary, but since the spec doesn’t govern stacks, i guess it didn’t come up.

Either way, for eslint’s purposes i don’t think it’s appropriate to lump them in with either blocks or functions - they’re a distinct new kind of thing.

**Comment by nzakas** [comment]
> Either way, for eslint’s purposes i don’t think it’s appropriate to lump them in with either blocks or functions - they’re a distinct new kind of thing.

I agree with @ljharb here. I don’t think it’s appropriate to add to an existing rule. If we want to add a new rule, I think that would be most appropriate.

**Comment by mdjermanovic** [comment]
Makes sense. I'll remove that task from the list for the first version, and we can discuss a new rule later.

**Comment by btmills** [comment]
All currently-known work to implement class static blocks just shipped in ESLint v8.3.0!
